### ISSUE TITLE
link colors (unvisited and visited)

### ISSUE THREAD

**Issue by kengdoj** [Section508Coordinators/ICTTestingBaseline#177]
Do baseline tests cover the colors used to convey visited/unvisited meaning and color contrasts between visited and unvisited links? Consider when the browser sets these colors.

**Comment by kengdoj** [comment]
[WCAG 2.1 Understanding 1.4.11](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html), under Active User Interface Components:
"This Success Criterion does not require that changes in color that differentiate between states of an individual component meet the 3:1 contrast ratio when they do not appear next to each other. For example, there is not a new requirement that visited links contrast with the default color..."

That addresses the visited link color to unvisited link color since they are not adjacent. Still need to address use of color and contrast of these visited/unvisited colors.

**Comment by cliffbert** [comment]
Might be useful to list the different considerations from our discussion:

1. Color: Providing information through color only?
2. Contrast: Distinguishing between non-link and link text
3. Contrast: Distinguishing between visited and unvisited links (adjacency might be horizontal, too, e.g. a list of links)
4. Which of these can we reasonably expect the developer to control?

**Comment by kengdoj** [comment]
According to https://www.softwarehow.com/change-color-visited-links/, only Firefox has a built-in way to change link colors. Chrome, Safari require an extension. Edge has no setting to change link colors.

On https://yourbusiness.azcentral.com/changing-color-clicked-hyperlink-google-chrome-20016.html, instructions are provided to create a style file for Chrome: 

> Google Chrome enables you to edit a custom style sheet which overrides the settings of the pages you visit. Using this style sheet, you can specify exactly what color you prefer for clicked hyperlinks.

> Step 6
Add the line "a:visited {color: yourcolor !important;}" (without the quotation marks) at the end of the file. Replace "yourcolor" with the color of your choosing. Use either the name of the color or the hexadecimal color code to specify the color of the visited link. For example, to change the color of visited links to bright red, use either "a:visited {color: bright red !important;}" or "a:visited {color: #00ff00!important;}."

These instructions did not work for me on my gfe. I am guessing this is out of date.

**Comment by kengdoj** [comment]
Using Note from https://www.w3.org/WAI/WCAG21/Techniques/failures/F73:

> NOTE
> Red and Pink are the same color (hue) but they have different lightness (which is not color ). So red and pink would pass the requirement for "not distinguished by color (hue) alone" since they differ by lightness (which is not color) - as long as the difference in lightness (contrast) is 3:1 or greater. For example, if surrounding text is RED and the link is PINK it would pass. Similarly a light green and a dark red differ BOTH by color AND by lightness so they would pass if the contrast (lightness) difference is 3:1 or greater before focus or pointing.
> 
> There is no requirement that links be identifiable by people who cannot perceive color if they are not perceivable for those with color vision. (e.g. if the links are hidden for everyone – as in a game or test).
> 
> If the non-color cue only happens when the mouse hovers over the link or when the link receives focus, it is still a failure.
> 
> If the link is a different color and bold it would not fail because the boldness is not color dependent.

If there is no change in color for visited/unvisited links, ignore. If there is a color diff, add a test to check 3:1 contrast between the link colors. @drewnielson: A test case of visited/hover/unvisited colors that pass would be helpful to show that this is a passable test. (https://designsystem.digital.gov/design-tokens/color/overview/)

AND add test to check link colors against their background in contrast test (4.5:1)

**Comment by kengdoj** [comment]
If link visited/unvisited color will be included in testing:

This Advisory Tip from [5. Changing Content](https://section508coordinators.github.io/ICTTestingBaseline/05Changing.html) says that the name of the link must be updated to reflect the change in meaning. So, a visited link would need to indicate that it's been visited in its name.

> Changing content might also include changes in color to convey information. If so, this test should check that the name is updated to reflect the meaning of the changed color.

**Comment by drewnielson** [comment]
We might want to revisit - if the purpose of the change in the accessible name is solely to provide AT with the same information (visited/unvisited) as the color, then AT already has access to that information. Default setting in NVDA, for instance is to report "visited/unvisited link" whenever announcing a link. So providing visited/unvisited within the accessible name would actually make the information redundant and probably annoying.

**Comment by kengdoj** [comment]
According to this [AT and links testing](https://a11ysupport.io/tests/tech__html__links__example4), accessibility support for indicating the visited state of a link is inconsistent. So, if the visited information is added to the name of the link, it may or may not be redundant depending on the AT. 

Implementation question:  The name of the link would have to change when the the link has a visited state in the browser history. How hard is this to do (changing a link name temporarily)? I imagine if it's a hassle to remediate, developers will just set all link states to the same color.

Is there a link state property that can be revealed for testing? If so, I think this would be acceptable to indicate the change instead of requiring it to be in the name. (ANDI does not currently indicate link state.)

[Default link styles](https://developer.mozilla.org/en-US/docs/Learn/Accessibility/CSS_and_JavaScript):

> The standard link conventions are underlined and a different color **(default: blue) in their standard state**, another color variation when the link has previously been **visited (default: purple)**, and yet another color when the link is **activated (default: red)**. In addition, the mouse pointer changes to a pointer icon when links are moused over, and the link receives a highlight when focused (e.g. via tabbing) or activated. 

Here's what I found using colour contrast analyzer, on Google search results page:

- Foreground:#1A0DAB (blue default link color)
- Background:#660099 (purple visited link color)
- The contrast ratio is: 1.2:1

**Comment by cliffbert** [comment]
> According to this AT and links testing, accessibility support for indicating the visited state of a link is inconsistent. So, if the visited information is added to the name of the link, it may or may not be redundant depending on the AT.

True, but the information is already programmatically available, so it seems like a shortcoming of the AT if it doesn't say so.

**Comment by kengdoj** [comment]
@cliffbert, see [definition of accessibility supported](https://www.w3.org/TR/UNDERSTANDING-WCAG20/conformance.html#uc-accessibility-support-head) and 

Conformance Requirement 4
4. Only Accessibility-Supported Ways of Using Technologies: Only accessibility-supported ways of using technologies are relied upon to satisfy the success criteria. Any information or functionality that is provided in a way that is not accessibility supported is also available in a way that is accessibility supported. (See Understanding accessibility support.)

**Comment by kengdoj** [comment]
From @michaelhortongsa: [Privacy and the :visited selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Privacy_and_the_:visited_selector)

> Before about 2010, the CSS :visited selector allowed websites to uncover a user's browsing history and figure out what sites the user had visited. This was done through window.getComputedStyle and other techniques. This process was quick to execute, and made it possible not only to determine where the user had been on the web, but could also be used to guess a lot of information about the user's identity.

> To mitigate this problem, browsers have limited the amount of information that can be obtained from visited links.

**Comment by kengdoj** [comment]
Group decision - hold for next Baseline update

@drewnielson, please add your discoveries to this conversation so that this can be picked up in a future update discussion.

**Comment by drewnielson** [comment]
Previous assumption was that, when color is used to identify a link, then it could be possible to use only a combination of color and contrast to distinguish:
* A link from surrounding text
* A visited link from an unvisited link
* A link on hover and/or an active link from both visited link and univisted link

***HOWEVER***, it is **NOT** possible to use contrast to distinguish between a visited link, unvisited link, AND surrounding text while ALSO maintaining sufficient contrast with the background. Some other method must be used in addition to contrast (e.g., text decoration, weight, outline, cursor type, etc.).

See demo file for reference: [Visited/Unvisited Link Colors](https://section508coordinators.github.io/BaselineTestPages/docs/testfiles/TF07/07-visitedlinkcolors.html)

**Comment by kengdoj** [comment]
Here's a [page](https://zellwk.com/blog/style-hover-focus-active-states/) on states and different options to **NOT** rely on color.

**Comment by bruce-usab** [comment]
FWIW, there is lots of conflicted hand-waving from w3c AG WG folks on this, since their is a strong prejudice that the default browser behavior should be minimally acceptable.

If you have not seen it, WebAIM has an easy tool to test link text color against both unlinked text and page background:
http://webaim.org/resources/linkcontrastchecker

But it all kind of falls apart when trying to account for hover and visited states.

**Comment by kengdoj** [comment]
Resolution: Baseline will not include link colors to indicate visited/unvisited.

[WCAG 2.2 Understanding SC 1.4.1: Use of Color](https://www.w3.org/WAI/WCAG22/Understanding/use-of-color) states 

"Where color alone distinguishes between visited and unvisited links, it does not result in a failure of this Success Criterion, even where the contrast between the two link colors is below 3:1."

> Note
> 
> Most user agents provide users with a color-only cue that a link has been previously activated by them ("visited"). However, several technical constraints result in authors having very limited control over these color-only indications of visited links. The technical constraints are as follows:
> 
> 1. User agents constrain the exposure of a link's visited state due to [privacy concerns](https://developer.mozilla.org/en-US/docs/Web/CSS/Privacy_and_the_:visited_selector). Author queries to user agents will indicate all links have not been visited.
> 
> 1. Any available information on the visited state of a link would be inaccurate since it is both user and browser-dependent. Even if an author could accurately get information on previously activated links by a certain user, the author would be constrained to the current browser's preserved history, and would be unable to determine if the user had visited the page using a different browser (or if the history was not preserved, either from cache clearing or use of private sessions).
> 
> 1. Authors can only use color to modify the :visited CSS pseudoclass style. The technology constrains any non-color styling. Due to palette limitations, an author cannot use color alone to achieve 3:1 contrast between link and non-link text as well as between visited and unvisited links while also achieving 4.5:1 contrast for all link and non-link text.
> 
> 1. Authors also cannot set the visited state of links. The anchor element does not include a "visited" attribute; therefore the author has no ability to alter the state through an attribute setting. As such, authors cannot achieve [1.3.1 Info and Relationships](https://www.w3.org/WAI/WCAG22/Understanding/info-and-relationships) or [4.1.2 Name, Role, Value](https://www.w3.org/WAI/WCAG22/Understanding/name-role-value) in regard to visited links.
> 
> For these reasons, setting or conveying a link's visited status is not an author responsibility. Where color alone distinguishes between visited and unvisited links, it does not result in a failure of this Success Criterion, even where the contrast between the two link colors is below 3:1. Note that authors must continue to ensure that all text links meet contrast minimums against the page background (SC 1.4.3).
